consensus/istanbul/backend/engine.go

```diff
@@ -272,29 +270,103 @@ func (sb *Backend) verifyCommittedSeals(chain consensus.ChainReader, header *typ
 		return errEmptyCommittedSeals
 	}
 
+	// Check the signatures on the current header
+	snap, err := sb.snapshot(chain, header.Number.Uint64()-1, header.ParentHash, parents)
```

Comment:
You probably noticed this, but just want to highlight here that we call snapshot on the previous block here, because snapshot(blockN) gives us the validator set that needs to sign blockN+1

Reply:
Yeah that's why I write the `-1` and `ParentHash`, IMO the whole codebase needs to be refactored to not require that, it's very confusing